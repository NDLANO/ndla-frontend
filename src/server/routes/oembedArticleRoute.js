/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { BAD_REQUEST, INTERNAL_SERVER_ERROR } from 'http-status';
import { getArticleIdFromResource } from '../../containers/Resources/resourceHelpers';
import {
  fetchResource,
  fetchTopic,
} from '../../containers/Resources/resourceApi';
import config from '../../config';
import handleError from '../../util/handleError';
import { fetchArticle } from '../../containers/ArticlePage/articleApi';
import { parseAndMatchUrl } from '../../util/urlHelper';

function getOembedObject(req, title, html) {
  return {
    data: {
      type: 'rich',
      version: '1.0', // oEmbed version
      height: req.query.height ? req.query.height : 800,
      width: req.query.width ? req.query.width : 800,
      title,
      html,
    },
  };
}

const getHTMLandTitle = async match => {
  const {
    params: { resourceId, topicId, lang = 'nb' },
  } = match;
  if (!topicId && !resourceId) {
    return {};
  }
  if (topicId && !resourceId) {
    const topic = await fetchTopic(`urn:${topicId}`, lang);
    const articleId = getArticleIdFromResource(topic);
    return {
      title: topic.name,
      html: `<iframe aria-label="${topic.name}" src="${config.ndlaFrontendDomain}/article-iframe/${topic.id}/${articleId}?removeRelatedContent=true" frameborder="0" allowFullscreen="" />`,
    };
  }

  const resource = await fetchResource(`urn:resource:${resourceId}`, lang);
  const articleId = getArticleIdFromResource(resource);
  return {
    title: resource.name,
    html: `<iframe aria-label="${resource.name}" src="${config.ndlaFrontendDomain}/article-iframe/${resource.id}/${articleId}?removeRelatedContent=true" frameborder="0" allowFullscreen="" />`,
  };
};

export async function oembedArticleRoute(req) {
  const { url } = req.query;
  if (!url) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Missing url param.',
    };
  }

  const match = parseAndMatchUrl(url);
  if (!match) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Invalid url.',
    };
  }

  const {
    params: { resourceId, topicId, lang = 'nb' },
  } = match;
  try {
    if (!resourceId && !topicId) {
      const {
        params: { articleId },
      } = match;
      const article = await fetchArticle(articleId, lang);
      return getOembedObject(
        req,
        article.title,
        `<iframe aria-label="${article.title}" src="${config.ndlaFrontendDomain}/article-iframe/${lang}/article/${articleId}?removeRelatedContent=true" frameborder="0" allowFullscreen="" />`,
      );
    }
    const { html, title } = await getHTMLandTitle(match);
    return getOembedObject(req, title, html);
  } catch (error) {
    handleError(error);
    const status = error.status || INTERNAL_SERVER_ERROR;
    return {
      status,
      data: 'Internal server error',
    };
  }
}
