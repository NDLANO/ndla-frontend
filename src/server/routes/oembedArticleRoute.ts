/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { BAD_REQUEST, INTERNAL_SERVER_ERROR } from 'http-status';
import express from 'express';
import { match as RouterMatchType } from 'react-router';
import { getArticleIdFromResource } from '../../containers/Resources/resourceHelpers';
import {
  fetchResource,
  fetchTopic,
} from '../../containers/Resources/resourceApi';
import config from '../../config';
import handleError from '../../util/handleError';
import { fetchArticle } from '../../containers/ArticlePage/articleApi';
import { parseAndMatchUrl } from '../../util/urlHelper';

function getOembedObject(req: express.Request, title?: string, html?: string) {
  return {
    data: {
      type: 'rich',
      version: '1.0', // oEmbed version
      height: req.query.height ? req.query.height : 800,
      width: req.query.width ? req.query.width : 800,
      title,
      html,
    },
  };
}

interface MatchParams {
  resourceId?: string;
  topicId?: string;
  lang?: string;
  articleId?: string;
}
const getHTMLandTitle = async (match: RouterMatchType<MatchParams>) => {
  const {
    params: { resourceId, topicId, lang = 'nb' },
  } = match;
  if (!topicId && !resourceId) {
    return {};
  }
  if (topicId && !resourceId) {
    const topic = await fetchTopic(`urn:${topicId}`, lang);
    const articleId = getArticleIdFromResource(topic);
    return {
      title: topic.name,
      html: `<iframe aria-label="${topic.name}" src="${config.ndlaFrontendDomain}/article-iframe/${lang}/${topic.id}/${articleId}" frameborder="0" allowFullscreen="" />`,
    };
  }

  const resource = await fetchResource(`urn:resource:${resourceId}`, lang);
  const articleId = getArticleIdFromResource(resource);
  return {
    title: resource.name,
    html: `<iframe aria-label="${resource.name}" src="${config.ndlaFrontendDomain}/article-iframe/${lang}/${resource.id}/${articleId}" frameborder="0" allowFullscreen="" />`,
  };
};

export async function oembedArticleRoute(req: express.Request) {
  const { url } = req.query;
  if (!url || typeof url !== 'string') {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Missing url param.',
    };
  }

  const match = parseAndMatchUrl<MatchParams>(url);
  if (!match) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Invalid url.',
    };
  }

  const {
    params: { resourceId, topicId, lang = 'nb' },
  } = match;
  try {
    if (!resourceId && !topicId) {
      const {
        params: { articleId },
      } = match;
      const article = await fetchArticle(articleId, lang);
      return getOembedObject(
        req,
        article.title,
        `<iframe aria-label="${article.title}" src="${config.ndlaFrontendDomain}/article-iframe/${lang}/article/${articleId}" frameborder="0" allowFullscreen="" />`,
      );
    }
    const { html, title } = await getHTMLandTitle(match);
    return getOembedObject(req, title, html);
  } catch (error) {
    handleError(error);
    const status = error.status || INTERNAL_SERVER_ERROR;
    return {
      status,
      data: 'Internal server error',
    };
  }
}
