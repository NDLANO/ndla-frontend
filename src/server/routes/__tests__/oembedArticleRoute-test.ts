/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Request } from "express";
import nock from "nock";
import { oembedArticleRoute } from "../oembedArticleRoute";

const validArticleUrl1 = "https://test.ndla.no/subject:3/topic:1:55163/topic:1:168398/resource:1:1682";
const validArticleUrl2 = "https://test.ndla.no/subjects/subject:3/topic:1:55163/topic:1:168398/resource:1:1682";
const unvalidArticleUrl = "https://test.ndla.no/subject:3";

test("oembedArticleRoute success", async () => {
  nock("http://ndla-api").get("/taxonomy/v1/nodes/urn:resource:1:1682?language=nb").times(2).reply(200, {
    id: "urn:resource:1",
    contentUri: "urn:article:123",
    name: "Resource title",
  });

  const response = await oembedArticleRoute({
    query: {
      url: validArticleUrl1,
    },
  } as any as Request);

  expect(response).toMatchSnapshot();

  const response2 = await oembedArticleRoute({
    query: {
      url: validArticleUrl2,
    },
  } as any as Request);

  expect(response2).toMatchSnapshot();
  expect(nock.pendingMocks()).toStrictEqual([]);
});

test("oembedArticleRoute invalid url", async () => {
  const response = await oembedArticleRoute({
    query: {
      url: unvalidArticleUrl,
    },
  } as any as Request);

  expect(response).toMatchSnapshot();
});
