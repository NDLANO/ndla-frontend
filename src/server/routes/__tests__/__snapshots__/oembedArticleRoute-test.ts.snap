// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`oembedArticleRoute invalid url 1`] = `
{
  "data": "Bad request. Url not recognized",
  "status": 404,
}
`;

exports[`oembedArticleRoute learningpath fail 1`] = `
{
  "data": "Not found",
  "status": 404,
}
`;

exports[`oembedArticleRoute success 1`] = `
{
  "data": {
    "breadcrumbs": [
      "Fag",
      "Emne",
    ],
    "height": 480,
    "html": "<iframe aria-label="Resource title" src="https://test.ndla.no/article-iframe/nb/urn:resource:1/123" height="480" width="854" frameborder="0" allowFullscreen="" />",
    "iframeSrc": "https://test.ndla.no/article-iframe/nb/urn:resource:1/123",
    "resourceTypes": [
      {
        "connectionId": "",
        "id": "urn:resourcetype:subjectMaterial",
        "name": "Fagstoff",
        "parentId": null,
        "supportedLanguages": [],
        "translations": [],
      },
      {
        "connectionId": "",
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
        "parentId": "urn:resourcetype:subjectMaterial",
        "supportedLanguages": [],
        "translations": [],
      },
    ],
    "title": "Resource title",
    "type": "rich",
    "version": "1.0",
    "width": 854,
  },
  "status": 200,
}
`;

exports[`oembedArticleRoute success 2`] = `
{
  "data": {
    "breadcrumbs": [
      "Fag",
      "Emne",
    ],
    "height": 480,
    "html": "<iframe aria-label="Resource title" src="https://test.ndla.no/article-iframe/nb/urn:resource:1/123" height="480" width="854" frameborder="0" allowFullscreen="" />",
    "iframeSrc": "https://test.ndla.no/article-iframe/nb/urn:resource:1/123",
    "resourceTypes": [
      {
        "connectionId": "",
        "id": "urn:resourcetype:subjectMaterial",
        "name": "Fagstoff",
        "parentId": null,
        "supportedLanguages": [],
        "translations": [],
      },
      {
        "connectionId": "",
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
        "parentId": "urn:resourcetype:subjectMaterial",
        "supportedLanguages": [],
        "translations": [],
      },
    ],
    "title": "Resource title",
    "type": "rich",
    "version": "1.0",
    "width": 854,
  },
  "status": 200,
}
`;
